Removes deleted site from displaying

SowmiaS opened the pull request:
Fixes #[15592](https://github.com/wordpress-mobile/WordPress-Android/issues/15592)

## Now
Once the site is deleted, it should be reflected immediately in the app.
Now, the deleted site is still displayed until the app restarts.

## Description
So this PR fixes this issue by reflecting the deleted site immediately in the app.

### Before 
https://user-images.githubusercontent.com/4975407/157210184-205888f5-0f4d-49a8-b35c-e5d579f42d4a.mp4
### After
https://user-images.githubusercontent.com/4975407/157209770-e8ba4565-2b05-4579-8c78-7b2daf5ea237.mp4


## Test
#### Test 1
1. when there are **more than two sites**, delete a site.
2. Then it should delete the site
3. And Then navigate to the `Choose site` screen.
#### Test 2
1. when there are **two sites**, delete a site.
2. Then it should delete the site
3. And Then navigate to the `My Site` screen with the remaining site being selected automatically.
#### Test 3
1. when there are **one sites**, delete a site.
2. Then it should delete the site
3. And Then navigate to the `Create a new site` screen with the remaining site being selected automatically.

#### Test 3
1. when the user is neither signed in to a WordPress.com account nor if he has a .org site.
2. And deletes the site
4. And Then navigate to the `Signin` screen.

## Regression Notes
1. No Potential unintended areas of impact.
2. Adds unit test for the changes in WPMainActivityViewModel where possible.

PR submission checklist:

- [X] I have completed the Regression Notes.
- [X] I have considered adding accessibility improvements for my changes.
- [X] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

SowmiaS committed:
Removes selected site from Repository on site deletion

SowmiaS committed:
Displays SitePicker screen on site deletion when multiple sites available

SowmiaS committed:
Injects SelectedSiteRepository,AccountStore, SiteStore into WPMainActivityViewModel

SowmiaS committed:
WPMainActivityViewModel computes 'isMultipleSiteAvailable' property

SowmiaS committed:
Replaces use of mSiteStore in handleSiteRemoved function in  Activity with viewmodel properties

SowmiaS committed:
Replaces use of FluxCUtils with viewmodel property-isSignedInWPComOrHasWPOrgSite

SowmiaS committed:
Creates extension function showSignIn

SowmiaS committed:
Adds firstSite and isSignedInWPComOrHasWPOrgSite property to WPMainActivityViewModel

SowmiaS committed:
Replaces redundant code with AndroidExtensionsKt.showSignInForResult function

SowmiaS committed:
Removes nested if else statement

SowmiaS committed:
Replaces use of mSelectedSiteRepository in handleSiteRemoved function in activity with viewmodel handling

SowmiaS committed:
Adds empty check for firstSite

SowmiaS committed:
Corrects spacing issues warned by detekt

SowmiaS committed:
Adds parameters to WPMainActivityViewModel instance in Test file

SowmiaS committed:
Tests 'viewModel.isMultipleSiteAvailable' in WPMainActivityViewModelTest

SowmiaS committed:
Tests 'viewModel.firstSite' in WPMainActivityViewModelTest

AjeshRPai commented on .idea/inspectionProfiles/Project_Default.xml:
```diff
@@ -4,7 +4,7 @@
       <scope name="Tests" level="INFORMATION" enabled="true" />
     </inspection_tool>
-    <inspection_tool class="EnumSwitchStatementWhichMissesCases" enabled="true" level="ERROR" enabled_by_default="true">
+    <inspection_tool class="EnumSwitchStatementWhichMissesCases" enabled="false" level="ERROR" enabled_by_default="false">
```
@SowmiaS Please add these files to your git ignore local file. We can exclude these changes from the PR as these are only required for your local env

SowmiaS replied:
This was committed by mistake. Reverted back the changes done in inspectionprofiles. Done in commit https://github.com/wordpress-mobile/WordPress-Android/pull/16063/commits/f0de81d79fc75efdd570528ae43bf8fe94935edc

AjeshRPai replied:
@SowmiaS I think there is still some change in the file which is showing up in the  `Files changed` tab of the PR. Please have a look at it.

SowmiaS replied:
Sorry, It was my mistake. I have removed the trailing newline from `inspectionProfiles/Project_Default.xml`. Done in commit d3cc7a6e55a3ca45751e8553a7eead1d47a9f752

AjeshRPai commented on WordPress/src/main/java/org/wordpress/android/extension/AndroidExtensions.kt:
```diff
@@ -0,0 +1,13 @@
+        ActivityLauncher.showSignInForResultJetpackOnly(this)
+    } else {
+        ActivityLauncher.showSignInForResultWpComOnly(this)
+    }
```
@SowmiaS 👍 for extracting the `ActivityLauncher.showSignInForResultJetpackOnly` and `ActivityLauncher.showSignInForResult` to a method. 🙌 .I think we can move this function to the `WPMainActivity`. I think that `showSignInForResult` is not an extension required for `FragmentActivity`. We can move this as a normal function to the `WPMainActivity`. Wdyt?

SowmiaS replied:
@AjeshRPai It makes sense 🙂 . I have moved the function to `ActivityLauncher`, so that related code will be in one class. Done in commits - https://github.com/wordpress-mobile/WordPress-Android/pull/16063/commits/352e063c62c610fa98aa17dd2e117f981356fb75, https://github.com/wordpress-mobile/WordPress-Android/pull/16063/commits/47477c0359b67045f7701425132afe039ad94e40

AjeshRPai replied:
Suggestion 💡   @SowmiaS I think its better to have the method `showSignInForResultBasedOnIsJetpackAppBuildConfig` in `WPMainActivity` as this is only being used in the `WPMainActivity`. In addition to that, I feel the `ActivityLauncher` class should be responsible only to launch activity and not hold any logic. WDYT?. 🤔

SowmiaS replied:
Okay @AjeshRPai 🙂  . I have moved the method `showSignInForResultBasedOnIsJetpackAppBuildConfig` from `ActivityLauncher`  to `WPMainActivity`. Done in commit 9da995b205af10f470739fbfe3920df5f47716fe

AjeshRPai commented on WordPress/src/main/java/org/wordpress/android/ui/main/WPMainActivity.java:
```diff
@@ -1354,19 +1350,14 @@ private void updateConnectionBar(boolean isConnected) {
+        }
+        if (mViewModel.isMultipleSiteAvailable()) {
+            ActivityLauncher.showSitePickerForResult(this, mViewModel.getFirstSite());
+            return;
```
This `return` statement is redundant as this is the last statement.

SowmiaS replied:
I have removed the redundant return statement in [commit](https://github.com/wordpress-mobile/WordPress-Android/pull/16063/commits/643b8dec6a76175a2964cdfdc6294e0ec9551aa3).

AjeshRPai commented on WordPress/src/main/java/org/wordpress/android/viewmodel/main/WPMainActivityViewModel.kt:
```diff
@@ -38,14 +42,18 @@ import javax.inject.Inject
+    @Named(UI_THREAD) private val mainDispatcher: CoroutineDispatcher,
+    private val mSelectedSiteRepository: SelectedSiteRepository,
+    private val mAccountStore: AccountStore,
+    private val mSiteStore: SiteStore,
```
@SowmiaS 👍  on adding these dependencies to the `WPMainActivityViewModel` from the Activity. We can rename these variables without the 'm' prefix as that was a convention followed in the java code. for eg: `mSiteStore` can be renamed to `siteStore`. wdyt? 🤔

SowmiaS replied:
I agree @AjeshRPai. Done in commit https://github.com/wordpress-mobile/WordPress-Android/pull/16063/commits/ec02f37cda30b9213a1bfb3c563fff4f0869a0c5.

AjeshRPai commented on WordPress/src/test/java/org/wordpress/android/viewmodel/main/WPMainActivityViewModelTest.kt:
```diff
@@ -54,6 +57,9 @@ class WPMainActivityViewModelTest : BaseUnitTest() {
     @Mock lateinit var quickStartRepository: QuickStartRepository
+    @Mock lateinit var selectedSiteRepository: SelectedSiteRepository
+    @Mock lateinit var mAccountStore: AccountStore
+    @Mock lateinit var mSiteStore: SiteStore
```
@SowmiaS  We can rename these variables without the 'm' prefix as well. 🤔

SowmiaS replied:
I agree @AjeshRPai. Done in commit https://github.com/wordpress-mobile/WordPress-Android/pull/16063/commits/ec02f37cda30b9213a1bfb3c563fff4f0869a0c5.

AjeshRPai commented on WordPress/src/main/java/org/wordpress/android/viewmodel/main/WPMainActivityViewModel.kt:
```diff
@@ -99,6 +107,15 @@ class WPMainActivityViewModel @Inject constructor(
             .distinctUntilChanged()
             .map { Event(it) } as LiveData<Event<List<FocusPointInfo>>>
 
+    val isMultipleSiteAvailable: Boolean
```
Can be renamed to `hasMultipleSites`. Wdyt?. 🤔

SowmiaS replied:
It makes sense @AjeshRPai 🙂 . I have renamed to `hasMultipleSites`. Done in commit https://github.com/wordpress-mobile/WordPress-Android/pull/16063/commits/8bcb6feea611ad7bc7154d9dd89b72c73efe2c3b.

AjeshRPai commented on WordPress/src/test/java/org/wordpress/android/viewmodel/main/WPMainActivityViewModelTest.kt:
```diff
@@ -655,6 +664,43 @@ class WPMainActivityViewModelTest : BaseUnitTest() {
+
+    @Test
+    fun `firstSite should be return the first site available in the list of sites`() {
+        val sites = mock<ArrayList<SiteModel>>()
```
@SowmiaS I think an import was missing in this file for the `mock` helper function, Please make sure to push it as well. 
import statement - `import com.nhaarman.mockitokotlin2.mock`

SowmiaS replied:
@AjeshRPai. I have added the missing import in commit https://github.com/wordpress-mobile/WordPress-Android/pull/16063/commits/0b8d8798b81324cf3ef9972c49c7c9ae904745ef

AjeshRPai commented on WordPress/src/main/java/org/wordpress/android/viewmodel/main/WPMainActivityViewModel.kt:
```diff
@@ -99,6 +107,15 @@ class WPMainActivityViewModel @Inject constructor(
+    val isMultipleSiteAvailable: Boolean
+        get() = mSiteStore.sitesCount > ONE_SITE
+
+    val firstSite: SiteModel?
```
As the site returned by `firstSite` variable is the one selected by default in the site picker screen, can we rename this to `selectedSite`?. WDYT? 🤔

SowmiaS replied:
At this point of time, there are actually no selected sites. So naming the property as first site makes it clear that we are sending first site, which will be considered as selected site in site picker activity.
```kotlin
ActivityLauncher.showSitePickerForResult(this, mViewModel.getFirstSite()); // It seems clear that we are sending first site as selected site
ActivityLauncher.showSitePickerForResult(this, mViewModel.getSelectedSite());
```
@AjeshRPai If you still feel, it makes more sense when it is changed as `selectedSite`, I can do it. WDYT? 🤔

AjeshRPai replied:
>At this point of time, there are actually no selected sites. So naming the property as first site makes it clear that we are sending first site, which will be considered as selected site in site picker activity.

@SowmiaS 👋 I think you are right. Let's keep the naming as `firstSite` itself. 🙌

AjeshRPai commented on WordPress/src/test/java/org/wordpress/android/viewmodel/main/WPMainActivityViewModelTest.kt:
```diff
@@ -655,6 +664,43 @@ class WPMainActivityViewModelTest : BaseUnitTest() {
+    }
+
+    @Test
+    fun `firstSite should be return the first site available in the list of sites`() {
```
We can remove the `be` from the test statements. 

`firstSite should be return the first site available in the list of sites` can be changed to 
`selectedSite should return the first site available in the list of sites`

SowmiaS replied:
I have corrected the test function name as below
- firstSite should return null when there are no sites
- firstSite should return the first site available in the list of sites
 Done in commit https://github.com/wordpress-mobile/WordPress-Android/pull/16063/commits/03949a4d8ee0977530001486f2e4ae43a49c6c2c.

AjeshRPai reviewed (changes requested):
@SowmiaS Great work. 🙌 

I have tested the changes and everything looks good to me. I have tested for all the scenarios described in the parent issue and in the test block of the PR description. Everything works as Intended. I have left some minor comments and suggestions. 
Hope it helps, Let me know if anything is not clear.

SowmiaS committed:
Adds missing import com.nhaarman.mockitokotlin2.mock

SowmiaS committed:
Removes redundant return statement

SowmiaS committed:
Refactors the variables without m prefix

SowmiaS committed:
Refactors the variable name from  isMultipleSitesAvailable to hasMultipleSites

SowmiaS committed:
Refactors the test name in WPMainActivityViewModelTest
	- firstSite should return null when there are no sites
	- firstSite should return the first site available in the list of sites

SowmiaS committed:
Reverts inspectionProfile change
- Reverts EnumSwitchStatementWhichMissesCases back to true

SowmiaS committed:
Moves extension function showSignInForResult as showSignInForResultBasedOnIsJetpackAppBuildConfig static function in ActivityLauncher

SowmiaS committed:
Deletes unused extension folder

SowmiaS committed:
Fixes warning unnecessary mock in WPMainActivityViewModelTest

SowmiaS commented:
Thanks for taking your time to review this PR @AjeshRPai 🙇 . 

I agree with all the review comments and I have fixed all the comments given by you except **renaming `firstSite` to `selectedSite`**. 

> At this point of time, there are actually no selected sites. So naming the property as first site makes it clear that we are sending first site, which will be considered as selected site in site picker activity.
> 
```kotlin 
ActivityLauncher.showSitePickerForResult(this, mViewModel.getFirstSite()); // It seems clear that we are sending first site as selected site
ActivityLauncher.showSitePickerForResult(this, mViewModel.getSelectedSite());
```

@AjeshRPai If you still feel, it makes more sense when it is changed as `selectedSite`, I can do it. WDYT? 🤔

Kindly have a look at my changes.

AjeshRPai commented:
@SowmiaS, Thanks for making the changes 👍 . I am working on some other tasks which have a higher priority at the moment. I will take a look at these changes/ or assign someone as soon as possible. In the meanwhile may I suggest that you   update the PR by merging the [wordpress-mobile:trunk](https://github.com/wordpress-mobile/WordPress-Android)

SowmiaS committed:
Merge branch 'trunk' into issue/15592_deleted_site_still_visible_until_restart

SowmiaS committed:
Merge branch 'trunk' of ssh://github.com/SowmiaS/WordPress-Android into issue/15592_deleted_site_still_visible_until_restart

AliSoftware commented:
Note: don't forget to add a line to mention your fix in `RELEASE-NOTES.txt` before it gets merged 🙂

AjeshRPai commented:
Thanks, @AliSoftware for pointing it out. 🙌 

@SowmiaS Thanks for making the changes. I have added a minor comment and a suggestion. Please have a look at it. 

As @AliSoftware Pointed out, Could you add a line to mention your fix in ` RELEASE-NOTES.txt` as well?. You can find some examples in the same file.

AjeshRPai reviewed (changes requested):
Thanks, @AliSoftware for pointing it out. 🙌 

@SowmiaS Thanks for making the changes. I have added a minor comment and a suggestion. Please have a look at it. 

As @AliSoftware Pointed out, Could you add a line to mention your fix in ` RELEASE-NOTES.txt` as well?. You can find some examples in the same file.

SowmiaS committed:
Merge from trunk into issue/15592_deleted_site_still_visible_until_restart

SowmiaS committed:
Moves the function showSignInForResultBasedOnIsJetpackAppBuildConfig from AppLauncher to WPMainActivity

SowmiaS committed:
Removes trailing space from inspectionProfile

SowmiaS committed:
Updates RELEASE-NOTES.txt

SowmiaS committed:
Removes unused import

SowmiaS committed:
Replaces List.get(0) with indexing

SowmiaS commented:
@AjeshRPai Thanks for taking your time to review my changes. I have fixed all the review comments given by you.

> In the meanwhile may I suggest that you update the PR by merging the [wordpress-mobile:trunk](https://github.com/wordpress-mobile/WordPress-Android)

I have merged the latest code from [wordpress-mobile:trunk](https://github.com/wordpress-mobile/WordPress-Android)


> As @AliSoftware Pointed out, Could you add a line to mention your fix in RELEASE-NOTES.txt as well?. You can find some examples in the same file.

As pointed out by @AliSoftware and you, I have mentioned the fix in RELEASE-NOTES.txt.

AjeshRPai reviewed (commented):
I have looked at the code changes and tested it. Everything looks good to me. Great Job @SowmiaS . 👍 🙌 . Once again, Thank you for the contribution. I will merge this changes once the CI is done.

AjeshRPai commented on RELEASE-NOTES.txt:
```diff
@@ -10,6 +10,7 @@
 * [*] Block Editor: Add GIF badge for animated GIFs uploaded to Image blocks [https://github.com/WordPress/gutenberg/pull/38996]
 * [*] Block Editor: Small refinement to media upload errors, including centering and tweaking copy. [https://github.com/wordpress-mobile/gutenberg-mobile/pull/4597]
 * [*] Block Editor: Fix issue with list's starting index and the order [https://github.com/WordPress/gutenberg/pull/39354]
+* [*] My Site: Fixes an issue where a deleted site is displayed until the app restarts [https://github.com/wordpress-mobile/WordPress-Android/pull/16063]
```
@SowmiaS Hi 👋🏼 , Could you move the line describing your fix to the milestone - 19.6. Am sorry for not communicating it earlier or updating the milestone tag in PR.

SowmiaS replied:
No problem at all @AjeshRPai 🙂. I have updated the RELEASE-NOTES.txt, moving the release note related to this PR from milestone 19.5 to 19.6. Done in 6b5f5e96d95bbe11ee36dd54da23c30e451ef067

SowmiaS committed:
Updates RELEASE-NOTES.txt
- Moves the release note related to PR #16063 from 19.5 to 19.6 milestone

AliSoftware commented:
Good news @SowmiaS ! Your fix is now live and included in the new version (19.6) of WordPress Android that just shipped to the PlayStore 🎉 

Thanks again for your contribution!
